File: ardoc/ardoc_home_cleaner.pl
#!/usr/bin/env perl
use Env;
use File::Basename;
if ( "$#ARGV" < 0 || "$#ARGV" > 0 ){
print "ardoc_home_cleaner.pl: Error: one argument required:\n";
print "-----maximum allowed age of files/directories (days)\n";
exit 1;
} 

$days=$ARGV[0];
$lmt=$days*10;

my $ARDOC_HOME="$ARDOC_HOME";
my $ARDOC_WEBDIR="$ARDOC_WEBDIR";
my $ARDOC_JOB_LOG="$ARDOC_JOB_LOG";
if ( "$ARDOC_HOME" eq "" ){
    print "ardoc_home_cleaner.pl: Error: variable \$ARDOC_HOME is not defined \n";
    exit 1;
}
$ARDOC_HOME_DIR=dirname(${ARDOC_HOME});
if ( ! -d $ARDOC_HOME_DIR){
    print "ardoc_home_cleaner.pl: Error: area $ARDOC_HOME_DIR is not defined \n";
    exit 1;
}

print "ardoc_home_cleaner.pl: INFO: cleaning ardoc_local_gen_config files in $ARDOC_HOME_DIR \n";
opendir(WWD, ${ARDOC_HOME_DIR});
@allfiles = grep { !/^\.\.?\z/ }
readdir (WWD);
closedir(WWD);

@allfiles=sort{(stat("${area}/$a"))[9] <=> (stat("${area}/$b"))[9]} @allfiles;

foreach $ff (@allfiles){
    $mkm="${ARDOC_HOME_DIR}/${ff}";
    if ( $ff =~ /^ardoc_local_gen_config.*$/ ){
        $mkmk=`${ARDOC_HOME}/ardoc_mtime_diff.pl $mkm`;
        if ($mkmk <= $lmt) {last;}
        if ($mkmk > $lmt){ 
        print "ardoc_home_cleaner: age > ${days} day(s):  $ff $mkmk $lmt\n";
        print "ardoc_home_cleaner: DELETED:  $mkm\n";
        system("rm -rf ${mkm}");
        }
    }
}

if ( "$ARDOC_WEBDIR" ne "" && -d $ARDOC_WEBDIR){
    print "ardoc_home_cleaner.pl: INFO: cleaning web files areas in $ARDOC_WEBDIR \n";
    opendir(WWD, ${ARDOC_WEBDIR});
    @allfiles = grep { !/^\.\.?\z/ }
    readdir (WWD);
    closedir(WWD);

    @allfiles=sort{(stat("${area}/$a"))[9] <=> (stat("${area}/$b"))[9]} @allfiles;

    foreach $ff (@allfiles){
	$mkm="${ARDOC_WEBDIR}/${ff}";
	if ( $ff =~ /^ARDOC.*Log.*$/ ){
	    $mkmk=`${ARDOC_HOME}/ardoc_mtime_diff.pl $mkm`;
	    if ($mkmk <= $lmt) {last;}
	    if ($mkmk > $lmt){
		print "ardoc_home_cleaner: age > ${days} day(s):  $ff $mkmk $lmt\n";
		print "ardoc_home_cleaner: DELETED:  $mkm\n";
                system("rm -rf ${mkm}");
	    }
	}
    }
}

if ( "${ARDOC_JOB_LOG}" ne "" ){
    $ARDOC_JOB_LOGDIR=dirname(${ARDOC_JOB_LOG});
    if ( -d $ARDOC_JOB_LOGDIR){
        print "ardoc_home_cleaner.pl: INFO: cleaning job log area $ARDOC_JOB_LOGDIR\n";
	opendir(WWD, ${ARDOC_JOB_LOGDIR});
	@allfiles = grep { !/^\.\.?\z/ }
	readdir (WWD);
	closedir(WWD);

	@allfiles=sort{(stat("${area}/$a"))[9] <=> (stat("${area}/$b"))[9]} @allfiles;

	foreach $ff (@allfiles){
	    $mkm="${ARDOC_JOB_LOGDIR}/${ff}";
	    if ( $ff =~ /^log.*$/ ){
		$mkmk=`${ARDOC_HOME}/ardoc_mtime_diff.pl $mkm`;
		if ($mkmk <= $lmt) {last;}
		if ($mkmk > $lmt){
		    print "ardoc_home_cleaner: age > ${days} day(s):  $ff $mkmk $lmt\n";
		    print "ardoc_home_cleaner: DELETED:  $mkm\n";
                    system("rm -rf ${mkm}");
		}
	    }
	}
    }
}

File: ardoc/ardoc_home_cleaner.py
#!/usr/bin/env python3
import os
import sys
import re
import shutil
import time
from pathlib import Path

def main():
    if len(sys.argv) != 2:
        print("ardoc_home_cleaner.py: Error: one argument required:")
        print("-----maximum allowed age of files/directories (days)")
        sys.exit(1)

    try:
        days = int(sys.argv[1])
    except ValueError:
        print("ardoc_home_cleaner.py: Error: argument must be an integer.")
        sys.exit(1)

    now = time.time()
    age_in_seconds = days * 24 * 60 * 60

    ardoc_home = os.environ.get("ARDOC_HOME")
    if not ardoc_home:
        print("ardoc_home_cleaner.py: Error: variable $ARDOC_HOME is not defined")
        sys.exit(1)

    ardoc_home_dir = Path(ardoc_home).parent
    if not ardoc_home_dir.is_dir():
        print(f"ardoc_home_cleaner.py: Error: area {ardoc_home_dir} is not defined")
        sys.exit(1)

    def clean_directory(directory, pattern):
        print(f"ardoc_home_cleaner.py: INFO: cleaning files in {directory} matching '{pattern}'")
        for item in sorted(directory.iterdir(), key=lambda f: f.stat().st_mtime):
            if re.match(pattern, item.name):
                try:
                    item_age = now - item.stat().st_mtime
                    if item_age > age_in_seconds:
                        print(f"ardoc_home_cleaner: age > {days} day(s): {item.name}")
                        print(f"ardoc_home_cleaner: DELETED: {item}")
                        if item.is_dir():
                            shutil.rmtree(item)
                        else:
                            item.unlink()
                except FileNotFoundError:
                    # File might have been deleted by another process
                    continue


    clean_directory(ardoc_home_dir, r"^ardoc_local_gen_config.*$")

    ardoc_webdir = os.environ.get("ARDOC_WEBDIR")
    if ardoc_webdir and Path(ardoc_webdir).is_dir():
        clean_directory(Path(ardoc_webdir), r"^ARDOC.*Log.*$")

    ardoc_job_log = os.environ.get("ARDOC_JOB_LOG")
    if ardoc_job_log:
        ardoc_job_logdir = Path(ardoc_job_log).parent
        if ardoc_job_logdir.is_dir():
            clean_directory(ardoc_job_logdir, r"^log.*$")

if __name__ == "__main__":
    main()
